simple-mbed-cloud-client/simple-mbed-cloud-client.h

```diff
@@ -0,0 +1,208 @@
+
+#ifndef SIMPLEMBEDCLOUDCLIENT_H
+#define SIMPLEMBEDCLOUDCLIENT_H
+#include <stdio.h>
+#include "mbed-cloud-client/MbedCloudClient.h"
+#include "m2mdevice.h"
+#include "m2mresource.h"
+#include "mbed-client/m2minterface.h"
+#include "mbed-client/m2mvector.h"
+#include "mbed_cloud_client_resource.h"
+#include "mbed.h"
+#include "NetworkInterface.h"
+#include "storage_helper.h"
+
+class MbedCloudClientResource;
+
+class SimpleMbedCloudClient {
+
+public:
+
+    /**
+     * Initialize SimpleMbedCloudClient
+     *
+     * @param net A connected network interface
+     * @param bd An uninitialized block device
+     * @param fs An uninitialized file system
+     */
+    SimpleMbedCloudClient(NetworkInterface *net, BlockDevice *bd, FileSystem *fs);
+
+    /**
+     * SimpleMbedCloudClient destructor
+     *
+     * This deletes all managed resources
+     * This does not unregister the client
+     */
+    ~SimpleMbedCloudClient();
+
+    /**
+     * Initialize SimpleMbedCloudClient
+     *
+     * Sets up factory configurator client, loads root of trust,
+     * initializes storage, and loads existing developer credentials (if present)
+     *
+     * @param format If set to true, will always format the file system
+     *
+     * @returns 0 if successful, 1 if an error occured
+     */
+    int init(bool format = false);
+
+    /**
+     * Close the connection to Pelion Device Management and unregister the device
+     */
+    void close();
+
+    /**
+    * Sends a registration update message to the Cloud when the client is registered
+    * successfully to the Cloud and there is no internal connection error.
+    * If the client is not connected and there is some other internal network
+    * transaction ongoing, this function triggers an error MbedCloudClient::ConnectNotAllowed.
+    */
+    void register_update();
+
+    /**
+     * Checks registration status
+     *
+     * @returns true when connected, false when not connected
+     */
+    bool is_client_registered();
+
+    /**
+     * Whether the device has ever tried registering
+     *
+     * @returns true when registration was tried, false when not
+     */
+    bool is_register_called();
+
+    /**
+     * Register with Pelion Device Management
+     * This does *not* initialize the MbedCloudClientResource added through SimpleMbedCloudClient
+     * Only call this function after you unregistered first, do NOT use it to register the first time
+     *
+     * This function returns immediately, observe on_registered and on_error for updates.
+     *
+     * @returns true if connection started, false if connection was not started
+     */
+    bool call_register();
+
+    /**
+     * Create MbedCloudClientResource on the underlying Mbed Cloud Client object,
+     * register with Pelion Device Management.
+     *
+     * This function can only be called once. If you need to re-register,
+     * use `call_register`.
```

Comment:
I think this behavior should be improved, so there is no risk of calling the function twice.